mods/d2k/scripts/campaign.lua

```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+		end)
+		return
+	end
+	local actors = Map.ActorsInCircle(unit.CenterPosition, WDist.FromCells(5), function (a)
+		return
+		a.Owner.IsAlliedWith(bot) == false
+	end)
+	local targets = Utils.Where(actors, function(a)
+		return
+		a.Type == "light_inf" or
+		a.Type == "trooper" or
+		a.Type == "engineer" and
+		Map.TerrainType(a.Location) ~= "Rough"
 	end)
+	if targets[1] ~= nil then
+		unit.Stop()
+		unit.Move(Utils.Random(targets).Location)
+		Trigger.AfterDelay(55, function ()
+			AICrushLogic(unit, bot)
+		end)
+		unit.Hunt()
+	else
+		Trigger.AfterDelay(200, function ()
+			AICrushLogic(unit, bot)
+		end)
+	end
 end
+
+--- Select factories and unit types where crush logic should apply
+--- @param crusherTypes string[] list of units which use crush logic
+--- @param crusherFactories actor[] list of factories that produce crusher types
+function ActivateCrusherOnProductions(crusherTypes, crusherFactories)
+	Utils.Do(crusherFactories, function(factory)
+		Trigger.OnProduction(factory, function(producer, produced)
+			if not producer.Owner.IsBot then
+				return
+			end
+
+			local crusher = Utils.Any(crusherTypes, function(ct) return produced.Type == ct end)
+
+			if crusher then
+				AICrushLogic(produced, producer.Owner)
+			end
+		end)
+	end)
+end
+
+--- Periodically looks for new building build by BaseBuilder
+--- If production building is found, try restore production
+--- @param player player
+--- @param base actor[] list of all player buildings
+--- @param BuildingsToRebuild table<string, string[]> Key is Builidng type,
+--- @param delay fun(player):integer Function that returns a delay until production
+--- @param attackThresholdSize integer Function that returns a delay until production
+function FindNewBuilding(player, base, BuildingsToRebuild, delay, attackThresholdSize)
+	Trigger.AfterDelay(300, function()
+		for buildingType, productionTypes in pairs (BuildingsToRebuild) do
+			local buildings = player.GetActorsByType(buildingType)
+			for _, building in ipairs (buildings) do
+				if not Table_contains(base, building) then
+					Media.Debug("found new building"..building.Type)
+					table.insert(base,building)
+					DefendAndRepairBase(player, {building},0.5, AttackGroupSize[Difficulty] )
+					if #productionTypes > 0 then
+						Media.Debug("production type"..productionTypes[1])
+						local productionToBuild = function() return { Utils.Random(productionTypes) } end
+						ProduceUnits(player, building, delay, productionToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+					end
+				end
+			end
+		end
+		FindNewBuilding(player, base, BuildingsToRebuild, delay, attackThresholdSize)
+	end)
+end
+
+function IdleHuntOnBaseDestroyed(player, base)
+	local baseChecked = Utils.Where(base, function(building) return not building.IsDead end)
+	Trigger.OnAllKilledOrCaptured(baseChecked, function()
+		if Utils.Any(base, function(building) return not building.IsDead end)
+		then
+			IdleHuntOnBaseDestroyed(player, base)
```

Comment:
```suggestion
	local checkedBase = Utils.Where(base, function(building) return not building.IsDead and building.Owner == player end)
	Trigger.OnAllKilledOrCaptured(checkedBase, function()
		if Utils.Any(checkedBase, function(building) return not building.IsDead and building.Owner == player end)
		then
			IdleHuntOnBaseDestroyed(player, checkedBase)
```
Unless I'm misunderstanding the intent.